src/qibolab/_core/instruments/emulator/emulator.py

```diff
@@ -89,32 +89,45 @@ def _single_sequence(
         result for the execution of this single sequence, thus suitable
         to be returned as is.
         """
-        # probabilities for the |0>, |1> ... |n> state, for each swept value
+        # probabilities for states in computational basis
         probabilities = self._sweep(sequence, configs, sweepers)
         assert options.nshots is not None
-        # extract results from probabilities, according to the requested averaging mode
-        averaged = (
-            shots(probabilities, options.nshots)
-            if options.averaging_mode == AveragingMode.SINGLESHOT
-            # weighted averaged
-            else np.sum(probabilities * np.arange(probabilities.shape[-1]), axis=-1)
-        )
-        # move measurements dimension to the front, getting ready for extraction
-        res = np.moveaxis(averaged, -1, 0)
-
-        if options.acquisition_type is AcquisitionType.DISCRIMINATION:
-            measurements = res
-        elif options.acquisition_type is AcquisitionType.INTEGRATION:
-            x = np.stack((res, np.zeros_like(res)), axis=-1)
-            measurements = np.random.normal(x, scale=0.001)
-        else:
-            raise ValueError(
-                f"Acquisition type '{options.acquisition_type}' unsupported"
+        results = {}
+        levels = np.array(configs["hamiltonian"].transmon_levels)
+        for ro_id in self._acquisitions(sequence):
+            # TODO: find better to extract qubit from acquisition channel
+            # possible solution _single_sequence should be aware of platform.qubits
+            qubit = int(sequence.pulse_channels(ro_id)[0][0])
+            single_qubit_probabilities = np.moveaxis(
+                np.moveaxis(probabilities, -1, 0)[
+                    qubit * levels : (qubit + 1) * levels
+                ],
+                0,
+                -1,
+            )
```

Comment:
No need to create a single huge loop.
If it's not familiar to you to imagine the whole operation as an array, you could do it in steps.

Notice that, after this part, the whole operation does not depend on `qubit` any longer, but only on the probabilities. In turn, only `qubit` depends on `ro_id`.
So, you could just construct the array of probabilities within the loop, stacking the various `single_qubit_probabilities` into a single `probabilities` array, and then just keep it as before.

In a second step, you could just construct an array of `qubit` elements, and use that one to slice the original `probabilities` in a single operation.